api/v1/categories/PkpCategoryCategoryController.php

```diff
@@ -0,0 +1,409 @@
+
+        // Get a category object to edit or create
+        if ($categoryId == null) {
+            $category = Repo::category()->dao->newDataObject();
+            $category->setContextId($context->getId());
+        } else {
+            $category = Repo::category()->get($categoryId, $context->getId());
+
+            if (!$category) {
+                return response()->json(__('api.404.resourceNotFound'), Response::HTTP_NOT_FOUND);
+            }
+
+            $parentId = $category->getParentId();
+        }
+
+        $params = $this->convertStringsToSchema(\PKP\services\PKPSchemaService::SCHEMA_CATEGORY, $illuminateRequest->input());
+        $params['contextId'] = $category->getContextId();
+        $errors = Repo::category()->validate($category, $params, $context);
+
+        if (!empty($errors)) {
+            return response()->json($errors, Response::HTTP_BAD_REQUEST);
+        }
+
+        // Set the editable properties of the category object
+        $category->setTitle($illuminateRequest->input('title'), null);
+        $category->setDescription($illuminateRequest->input('description'), null);
+        $category->setParentId($parentId);
+        $category->setPath($illuminateRequest->input('path'));
+        $category->setSortOption($illuminateRequest->input('sortOption'));
+
+        // Update or insert the category object
+        if ($categoryId == null) {
+            $category->setSequence(REALLY_BIG_NUMBER);
+            $categoryId = Repo::category()->add($category);
+            Repo::category()->dao->resequenceCategories($context->getId());
+        } else {
+            Repo::category()->edit($category, []);
+        }
+
+        // Update category editors
+        // Expects subEditor to be an array of sub-editor IDs, grouped by Group IDs.
+        $subEditors = $illuminateRequest->input('subEditors');
+
+        // Subeditors are assigned to user group that has access to WORKFLOW_STAGE_ID_SUBMISSION stage, but OPS does not have that stage.
+        // So $subEditors will be null in OPS.
+        if ($subEditors !== null) {
+            Repo::category()->updateEditors($categoryId, $subEditors, Category::$ASSIGNABLE_ROLES, $context->getId());
+        }
+
+        $submittedImageData = $illuminateRequest->input('image') ?: [];
+        $temporaryFileId = $submittedImageData ? $submittedImageData['temporaryFileId'] : null;
+
+        // Delete the old image if a new one was submitted or if the existing one was removed
+        if ($temporaryFileId || !$submittedImageData) {
+            $oldImageData = $category->getImage();
+            if ($oldImageData) {
+                $publicFileManager = new PublicFileManager();
+                $publicFileManager->removeContextFile($category->getContextId(), $oldImageData['uploadName']);
+                $publicFileManager->removeContextFile($category->getContextId(), $oldImageData['thumbnailName']);
+                $category->setImage(null);
+            }
+        }
```

Comment:
Just to decrease the indentation

```suggestion
        $oldImageData = $category->getImage();
        if (($temporaryFileId || !$submittedImageData) && $oldImageData) {
            $publicFileManager = new PublicFileManager();
            $publicFileManager->removeContextFile($category->getContextId(), $oldImageData['uploadName']);
            $publicFileManager->removeContextFile($category->getContextId(), $oldImageData['thumbnailName']);
            $category->setImage(null);
        }
```